frappe_graphql/utils/generate_sdl/doctype.py

```diff
@@ -0,0 +1,174 @@
+import inflect
+
+import frappe
+from frappe.utils import cint
+from frappe.model import default_fields, display_fieldtypes, table_fields
+from frappe.model.meta import Meta
+
+
+def get_doctype_sdl(doctype, ignore_custom_fields=False):
+    meta = frappe.get_meta(doctype)
+    sdl, defined_fieldnames = get_basic_doctype_sdl(meta)
+
+    # Extend Doctype with Custom Fields
+    if not ignore_custom_fields and len(meta.get_custom_fields()):
+        sdl += get_custom_field_sdl(meta, defined_fieldnames)
+
+    # DocTypeSortingInput
+    sdl += get_sorting_input(meta)
+
+    sdl += get_connection_type(meta)
+
+    # Extend QueryType
+    sdl += get_query_type_extension(meta)
+
+    return sdl
+
+
+def get_basic_doctype_sdl(meta: Meta):
+    dt = format_doctype(meta.name)
+    sdl = f"type {dt} implements BaseDocType {{"
+
+    defined_fieldnames = [] + list(default_fields)
+
+    for field in default_fields:
+        if field in ("idx", "docstatus"):
+            fieldtype = "Int"
+        elif field in ("owner", "modified_by"):
+            fieldtype = "User!"
+        else:
+            fieldtype = "String"
+        sdl += f"\n  {field}: {fieldtype}"
+    sdl += "\n  owner__name: String!"
+    sdl += "\n  modified_by__name: String!"
+
+    for field in meta.fields:
+        if field.fieldtype in display_fieldtypes:
+            continue
+        if field.fieldname in defined_fieldnames:
+            continue
+        if cint(field.get("is_custom_field")):
+            continue
+        defined_fieldnames.append(field.fieldname)
+        sdl += f"\n  {get_field_sdl(field)}"
+        if field.fieldtype == "Link":
+            sdl += f"\n  {get_link_field_name_sdl(field)}"
+
+    sdl += "\n}"
+
+    return sdl, defined_fieldnames
+
+
+def get_custom_field_sdl(meta, defined_fieldnames):
+    sdl = f"\n\nextend type {format_doctype(meta.name)} {{"
+    for field in meta.get_custom_fields():
+        if field.fieldtype in display_fieldtypes:
+            continue
+        if field.fieldname in defined_fieldnames:
+            continue
+        defined_fieldnames.append(field.fieldname)
+        sdl += f"\n  {get_field_sdl(field)}"
+        if field.fieldtype == "Link":
+            sdl += f"\n  {get_link_field_name_sdl(field)}"
+    sdl += "\n}"
+
+    return sdl
+
+
+def get_sorting_input(meta):
+    dt = format_doctype(meta.name)
+
+    sdl = f"\n\nenum {dt}SortField {{"
+    sdl += "\n  NAME"
+    sdl += "\n  CREATION"
+    sdl += "\n  MODIFIED"
+    for field in meta.fields:
+        if not field.search_index and not field.unique:
```

Comment:
@faztp12 I think the unique check is not always needed. For instance, product would be sortable by its name, while its name is not necessarily unique.

Reply:
@assemmarwan in database, unique data columns are indexed..
so, if a field is neither search_indexed nor unique, we do not let them sort by that column.

in the product's case, the product_name should have search_index: 1 so it will be sortable

Reply:
Yes `search_index` is fine, but my concern is that the condition requires both 😅 . Should it be `or`?
```suggestion
        if not field.search_index or not field.unique:
```

Reply:
if its neither of them, continue.. it should be `and` right ? 😄

Reply:
My bad, please ignore what I said, I misread it as `!(search_index and unique)`